group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1087,1348 @@ public void replay(
+            } catch (Throwable t) {
+                responseFuture.complete(new JoinGroupResponseData()
+                    .setMemberId(memberId)
+                    .setErrorCode(Errors.forException(t).code())
+                );
+                return EMPTY_RESULT;
+            }
+
+            CoordinatorResult<Void, Record> newGroupResult = EMPTY_RESULT;
+            if (group.isNew()) {
+                // If a group was newly created, we need to append records to the log
+                // to commit the group to the timeline datastructure. We set this here
+                // while the group is still empty.
+
+                CompletableFuture<Void> appendFuture = new CompletableFuture<>();
+                appendFuture.whenComplete((__, t) -> {
+                    if (t != null) {
+                        // We failed to write the empty group metadata. This will revert the snapshot, removing
+                        // the newly created group.
+                        log.warn("Failed to write empty metadata for group {}: {}", group.groupId(), t.getMessage());
+                    }
+                });
+
+                List<Record> records = Collections.singletonList(
+                    RecordHelpers.newGroupMetadataRecord(group, metadataImage.features().metadataVersion())
+                );
+
+                newGroupResult = new CoordinatorResult<>(records, appendFuture);
+            }
+
+            String joinReason = request.reason();
+            if (joinReason == null || joinReason.isEmpty()) {
+                joinReason = "not provided";
+            }
+
+            if (!acceptJoiningMember(group, memberId)) {
+                group.remove(memberId);
+                responseFuture.complete(new JoinGroupResponseData()
+                    .setMemberId(UNKNOWN_MEMBER_ID)
+                    .setErrorCode(Errors.GROUP_MAX_SIZE_REACHED.code())
+                );
+
+            } else if (isUnknownMember) {
+                result = genericGroupJoinNewMember(
+                    context,
+                    request,
+                    group,
+                    joinReason,
+                    responseFuture
+                );
+            } else {
+                result = genericGroupJoinExistingMember(
+                    context,
+                    request,
+                    group,
+                    joinReason,
+                    responseFuture
+                );
+            }
+
+            // Attempt to complete join group phase. We do not complete
+            // the join group phase if this is the initial rebalance.
+            if (group.isInState(PREPARING_REBALANCE) &&
+                group.hasAllMembersJoined() &&
+                group.generationId() != 0
+            ) {
+                CoordinatorResult<Void, Record> joinResult = completeGenericGroupJoin(group);
+                result = result != EMPTY_RESULT ? result : joinResult;
```

Comment:
I am not satisfied with this logic here. I think that other folks won't understand this... We need to come up with a better way. I will think about it.

I don't recall if I already asked this but would it be possible to push `completeGenericGroupJoin` into `genericGroupJoinNewMember` and `genericGroupJoinExistingMember` instead of having it here?